src/tasks/task_list.hpp

```diff
@@ -27,54 +30,249 @@
 #include "task_id.hpp"
 #include "task_types.hpp"
+#include "utils/error_checking.hpp"
 
 namespace parthenon {
 
 enum class TaskListStatus { running, stuck, complete, nothing_to_do };
 
+class TaskList;
+namespace task_list_impl {
+TaskID AddTaskHelper(TaskList *, Task);
+} // namespace task_list_impl
+
+class IterativeTasks {
+ public:
+  IterativeTasks(TaskList *tl, int key) : tl_(tl), key_(key) {
+    max_iterations_ = std::numeric_limits<unsigned int>::max();
+  }
+
+  template <class T, class... Args>
+  TaskID AddTask(TaskID const &dep, T &&func, Args &&... args) {
+    return AddTask_(TaskType::iterative, 1, dep, std::forward<T>(func),
+                    std::forward<Args>(args)...);
+  }
+  // overload to add member functions of class T to task list
+  // NOTE: we must capture the object pointer
+  template <class T, class... Args>
+  TaskID AddTask(TaskID const &dep, TaskStatus (T::*func)(Args...), T *obj,
+                 Args &&... args) {
+    return this->AddTask_(TaskType::iterative, 1, dep, [=]() mutable -> TaskStatus {
+      return (obj->*func)(std::forward<Args>(args)...);
+    });
+  }
+
+  template <class T, class... Args>
+  TaskID SetCompletionTask(TaskID const &dep, T &&func, Args &&... args) {
+    return AddTask_(TaskType::completion_criteria, check_interval_, dep,
+                    std::forward<T>(func), std::forward<Args>(args)...);
+  }
+  template <class T, class... Args>
+  TaskID SetCompletionTask(TaskID const &dep, TaskStatus (T::*func)(Args...), T *obj,
+                           Args &&... args) {
+    return AddTask_(TaskType::completion_criteria, check_interval_, dep,
+                    [=]() mutable -> TaskStatus {
+                      return (obj->*func)(std::forward<Args>(args)...);
+                    });
+  }
+
+  void SetMaxIterations(const unsigned int max) { max_iterations_ = max; }
+  void SetCheckInterval(const unsigned int chk) { check_interval_ = chk; }
+  void SetFailWithMaxIterations(const bool flag) { throw_with_max_iters_ = flag; }
+  void SetWarnWithMaxIterations(const bool flag) { warn_with_max_iters_ = flag; }
+  bool ShouldThrowWithMax() const { return throw_with_max_iters_; }
+  bool ShouldWarnWithMax() const { return warn_with_max_iters_; }
+  unsigned int GetMaxIterations() const { return max_iterations_; }
+  unsigned int GetIterationCount() const { return count_; }
+  void IncrementCount() { count_++; }
+
+ private:
+  template <class F, class... Args>
+  TaskID AddTask_(const TaskType &type, const int interval, TaskID const &dep, F &&func,
+                  Args &&... args) {
+    TaskID id(0);
+    id = task_list_impl::AddTaskHelper(
+        tl_, Task(
+                 id, dep,
+                 [=, func = std::forward<F>(func)]() mutable -> TaskStatus {
+                   return func(std::forward<Args>(args)...);
+                 },
+                 type, key_));
+    return id;
+  }
+  TaskList *tl_;
+  int key_;
+  unsigned int max_iterations_;
```

Comment:
It's always set in the constructor -- is that not sufficient?

Reply:
It is sufficient. This is just a stylistic point here, because all the other fields are having their default values set in this section. would just be nice to be consistent.